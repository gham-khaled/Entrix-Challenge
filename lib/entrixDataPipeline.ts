import {Construct} from 'constructs';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as cdk from 'aws-cdk-lib';
import * as tasks from 'aws-cdk-lib/aws-stepfunctions-tasks'
import * as sfn from 'aws-cdk-lib/aws-stepfunctions'
import * as s3 from 'aws-cdk-lib/aws-s3'
import * as sns from 'aws-cdk-lib/aws-sns'
import {Duration} from 'aws-cdk-lib';
import {
    EventbridgeToStepfunctions,
    EventbridgeToStepfunctionsProps
} from '@aws-solutions-constructs/aws-eventbridge-stepfunctions';
import * as stepfunctions from 'aws-cdk-lib/aws-stepfunctions';
import * as events from 'aws-cdk-lib/aws-events';

export class EntrixDataPipelineStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);
        // S3 Bucket
        const logBucket = new s3.Bucket(this, 'MyLogBucket');

        // Lambda Functions
        const lambdaA = new lambda.Function(this, 'lambdaA', {
            runtime: lambda.Runtime.PYTHON_3_9,
            handler: 'lambda_a.lambda_handler',
            code: lambda.Code.fromAsset('lib/lambda/dataPipeline/'),
        });
        const lambdaB = new lambda.Function(this, 'lambdaB', {
            runtime: lambda.Runtime.PYTHON_3_9,
            handler: 'lambda_b.lambda_handler',
            code: lambda.Code.fromAsset('lib/lambda/dataPipeline/'),
            environment: {
                LOG_BUCKET: logBucket.bucketName
            },
        });

        // SNS Topic
        const errorTopic = new sns.Topic(this, 'ErrorTopic');

        // SF Tasks
        const LambdaAJob = new tasks.LambdaInvoke(this, 'LambdaA', {
            lambdaFunction: lambdaA,
        });

        const publishError = new tasks.SnsPublish(this, 'Publish Error', {
            topic: errorTopic,
            message: sfn.TaskInput.fromJsonPathAt('$')
        });

        const LambdaBJob = new tasks.LambdaInvoke(this, 'LambdaB', {
            lambdaFunction: lambdaB,
        }).addCatch(publishError, {errors: [sfn.Errors.TASKS_FAILED]});

        const mapLambdaBJob = new sfn.Map(this, 'LambdaBMap', {
            itemsPath: sfn.JsonPath.stringAt('$.orders'),
        }).itemProcessor(LambdaBJob)

        const definition = LambdaAJob.next(new sfn.Choice(this, 'Results Empty?', {inputPath: "$.Payload"})
            .when(sfn.Condition.booleanEquals('$.results', false), LambdaAJob)
            .when(sfn.Condition.booleanEquals('$.results', true), mapLambdaBJob)
            .afterwards())


        // SF Cron
        const constructProps: EventbridgeToStepfunctionsProps = {
            stateMachineProps: {
                definition: definition
            },
            eventRuleProps: {
                schedule: events.Schedule.rate(Duration.hours(1))
            }
        };

        new EventbridgeToStepfunctions(this, 'hourlyDataPipeline', constructProps);

        // Permissions
        logBucket.grantPut(lambdaB);
        // errorTopic.addSubscription(new subscriptions.EmailSubscription('gham.khaled@gmail.com'));


    }

}