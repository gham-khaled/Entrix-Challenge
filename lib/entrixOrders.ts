import {Construct} from 'constructs';
import * as dynamodb from 'aws-cdk-lib/aws-dynamodb';
import * as lambda from 'aws-cdk-lib/aws-lambda';
import * as apigateway from 'aws-cdk-lib/aws-apigateway';
import * as cdk from 'aws-cdk-lib';

export class EntrixOrdersStack extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        // Dynamo DB Table
        const recordsTable = new dynamodb.Table(this, 'recordsTable', {
            partitionKey: {name: 'record_id', type: dynamodb.AttributeType.STRING},
            billingMode: dynamodb.BillingMode.PAY_PER_REQUEST,
        });

        // Lambda Function
        const postLambda = new lambda.Function(this, 'postLambda', {
            runtime: lambda.Runtime.PYTHON_3_9,
            handler: 'post_lambda.lambda_handler',
            code: lambda.Code.fromAsset('lib/lambda/recordsService'),
            environment: {
                DYNAMODB_TABLE_NAME: recordsTable.tableName,
                LOGLEVEL: "INFO",
            },
        });
        recordsTable.grantWriteData(postLambda)

        // API Gateway
        const api = new apigateway.RestApi(this, 'recordsAPI');


        const requestValidator = new apigateway.RequestValidator(this, 'BodyRequestValidator', {
            restApi: api,
            validateRequestBody: true,
        });

        const recordsRequestModel = api.addModel('RecordsRequestModel', {
            contentType: 'application/json',
            modelName: 'RecordsRequestModel',
            schema: {
                type: apigateway.JsonSchemaType.ARRAY,
                items: {
                    type: apigateway.JsonSchemaType.OBJECT,
                    properties: {
                        record_id: {type: apigateway.JsonSchemaType.STRING},
                    },
                    required: ['record_id'],
                },
            },
        });
        api.root.addMethod('POST',  new apigateway.LambdaIntegration(postLambda), {
            requestValidator: requestValidator,
            requestModels: {
                'application/json': recordsRequestModel,
            },
        });


    }

}
