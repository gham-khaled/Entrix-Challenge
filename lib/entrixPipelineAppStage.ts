import * as cdk from 'aws-cdk-lib';
import {Construct} from "constructs";
import {EntrixOrdersStack} from "./entrixOrders";
import {EntrixDataPipelineStack} from "./entrixDataPipeline";
import {CfnOutput} from "aws-cdk-lib";

export class EntrixPipelineAppStage extends cdk.Stage {
    public readonly urlOutput: CfnOutput;

    constructor(scope: Construct, id: string, props?: cdk.StageProps) {
        super(scope, id, props);
        const entrixOrders = new EntrixOrdersStack(this, 'EntrixOrders');
        new EntrixDataPipelineStack(this, 'EntrixOrders');
        this.urlOutput = entrixOrders.urlOutput;

    }
}