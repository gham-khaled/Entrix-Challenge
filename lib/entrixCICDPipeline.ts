import * as cdk from 'aws-cdk-lib';
import {Construct} from 'constructs';
import {CodePipeline, CodePipelineSource, ShellStep} from 'aws-cdk-lib/pipelines';
import {EntrixPipelineAppStage} from "./entrixPipelineAppStage";

export class EntrixCICDPipeline extends cdk.Stack {
    constructor(scope: Construct, id: string, props?: cdk.StackProps) {
        super(scope, id, props);

        const pipeline = new CodePipeline(this, 'Pipeline', {
            pipelineName: 'MyPipeline',
            synth: new ShellStep('Synth', {
                input: CodePipelineSource.gitHub('gham-khaled/Entrix-Challenge', 'master'),
                installCommands: ['npm ci', 'npm run build'],
                commands: ['npm run test', 'npx cdk synth']
            })
        });

        const devApp = new EntrixPipelineAppStage(this, "Dev", {env: {account: '830962405258', region: 'eu-west-1'}})

        const devStage = pipeline.addStage(devApp);

        devStage.addPost(new ShellStep("API Testing", {
            envFromCfnOutputs: {ENDPOINT_URL: devApp.urlOutput},
            commands: ['curl -Ssf  --location $ENDPOINT_URL \\\n' +
            '--header \'Content-Type: application/json\' \\\n' +
            '--data \'[\n' +
            ' {\n' +
            ' "record_id": "unique_id_1",\n' +
            ' "parameter_1": "abc",\n' +
            ' "parameter_2": 4\n' +
            ' },{\n' +
            ' "record_id": "unique_id_2",\n' +
            ' "parameter_1": "def",\n' +
            ' "parameter_2": 2.1\n' +
            ' }\n' +
            ']\'']

        }));

    }
}