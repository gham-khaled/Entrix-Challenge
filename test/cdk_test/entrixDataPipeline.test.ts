import {Capture, Template} from "aws-cdk-lib/assertions";
import * as cdk from 'aws-cdk-lib';
import {EntrixDataPipelineStack} from "../../lib/entrixDataPipeline";
import {EntrixOrdersStack} from "../../lib/entrixOrders";

// Test Resource Creation

test('Lambda Functions Created', () => {
    const stack = new cdk.Stack();
    const entrixDataPipelineStack = new EntrixDataPipelineStack(stack, 'EntrixDataPipeline');
    const template = Template.fromStack(entrixDataPipelineStack);
    template.resourceCountIs("AWS::Lambda::Function", 2);
});
test('S3 Bucket Created', () => {
    const stack = new cdk.Stack();
    const entrixDataPipelineStack = new EntrixDataPipelineStack(stack, 'EntrixDataPipeline');
    const template = Template.fromStack(entrixDataPipelineStack);
    template.resourceCountIs("AWS::S3::Bucket", 1);
});
test('SNS Topic Created', () => {
    const stack = new cdk.Stack();
    const entrixDataPipelineStack = new EntrixDataPipelineStack(stack, 'EntrixDataPipeline');
    const template = Template.fromStack(entrixDataPipelineStack);
    template.resourceCountIs("AWS::SNS::Topic", 1);
});
test('Step Functions Created', () => {
    const stack = new cdk.Stack();
    const entrixDataPipelineStack = new EntrixDataPipelineStack(stack, 'EntrixDataPipeline');
    const template = Template.fromStack(entrixDataPipelineStack);
    template.resourceCountIs("AWS::StepFunctions::StateMachine", 1);
});
test('Event Rule Created', () => {
    const stack = new cdk.Stack();
    const entrixDataPipelineStack = new EntrixDataPipelineStack(stack, 'EntrixDataPipeline');
    const template = Template.fromStack(entrixDataPipelineStack);
    template.resourceCountIs("AWS::Events::Rule", 1);
});

// Test Resource Creation
test('Lambda Has Environment Variables', () => {
    const stack = new cdk.Stack();
    const entrixDataPipelineStack = new EntrixDataPipelineStack(stack, 'EntrixDataPipeline');
    const template = Template.fromStack(entrixDataPipelineStack);
    const envCapture = new Capture();
    template.hasResourceProperties("AWS::Lambda::Function", {
        Handler: "lambda_b.lambda_handler",
        Environment: envCapture,
    });

    expect(envCapture.asObject()).toEqual(
        {
            Variables: {
                LOG_BUCKET: {
                    Ref: "MyLogBucketBD20E7D7",
                },
            },
        }
    );
});