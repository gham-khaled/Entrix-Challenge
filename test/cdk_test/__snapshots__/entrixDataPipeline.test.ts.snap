// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Testing 1`] = `
{
  "Mappings": {
    "ServiceprincipalMap": {
      "af-south-1": {
        "states": "states.af-south-1.amazonaws.com",
      },
      "ap-east-1": {
        "states": "states.ap-east-1.amazonaws.com",
      },
      "ap-northeast-1": {
        "states": "states.ap-northeast-1.amazonaws.com",
      },
      "ap-northeast-2": {
        "states": "states.ap-northeast-2.amazonaws.com",
      },
      "ap-northeast-3": {
        "states": "states.ap-northeast-3.amazonaws.com",
      },
      "ap-south-1": {
        "states": "states.ap-south-1.amazonaws.com",
      },
      "ap-south-2": {
        "states": "states.ap-south-2.amazonaws.com",
      },
      "ap-southeast-1": {
        "states": "states.ap-southeast-1.amazonaws.com",
      },
      "ap-southeast-2": {
        "states": "states.ap-southeast-2.amazonaws.com",
      },
      "ap-southeast-3": {
        "states": "states.ap-southeast-3.amazonaws.com",
      },
      "ap-southeast-4": {
        "states": "states.ap-southeast-4.amazonaws.com",
      },
      "ca-central-1": {
        "states": "states.ca-central-1.amazonaws.com",
      },
      "cn-north-1": {
        "states": "states.cn-north-1.amazonaws.com",
      },
      "cn-northwest-1": {
        "states": "states.cn-northwest-1.amazonaws.com",
      },
      "eu-central-1": {
        "states": "states.eu-central-1.amazonaws.com",
      },
      "eu-central-2": {
        "states": "states.eu-central-2.amazonaws.com",
      },
      "eu-north-1": {
        "states": "states.eu-north-1.amazonaws.com",
      },
      "eu-south-1": {
        "states": "states.eu-south-1.amazonaws.com",
      },
      "eu-south-2": {
        "states": "states.eu-south-2.amazonaws.com",
      },
      "eu-west-1": {
        "states": "states.eu-west-1.amazonaws.com",
      },
      "eu-west-2": {
        "states": "states.eu-west-2.amazonaws.com",
      },
      "eu-west-3": {
        "states": "states.eu-west-3.amazonaws.com",
      },
      "il-central-1": {
        "states": "states.il-central-1.amazonaws.com",
      },
      "me-central-1": {
        "states": "states.me-central-1.amazonaws.com",
      },
      "me-south-1": {
        "states": "states.me-south-1.amazonaws.com",
      },
      "sa-east-1": {
        "states": "states.sa-east-1.amazonaws.com",
      },
      "us-east-1": {
        "states": "states.us-east-1.amazonaws.com",
      },
      "us-east-2": {
        "states": "states.us-east-2.amazonaws.com",
      },
      "us-gov-east-1": {
        "states": "states.us-gov-east-1.amazonaws.com",
      },
      "us-gov-west-1": {
        "states": "states.us-gov-west-1.amazonaws.com",
      },
      "us-iso-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-iso-west-1": {
        "states": "states.amazonaws.com",
      },
      "us-isob-east-1": {
        "states": "states.amazonaws.com",
      },
      "us-west-1": {
        "states": "states.us-west-1.amazonaws.com",
      },
      "us-west-2": {
        "states": "states.us-west-2.amazonaws.com",
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ErrorTopicA0904A23": {
      "Type": "AWS::SNS::Topic",
    },
    "MyLogBucketBD20E7D7": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
    "hourlyDataPipelineEventsRule9EBB6664": {
      "Properties": {
        "ScheduleExpression": "rate(1 hour)",
        "State": "ENABLED",
        "Targets": [
          {
            "Arn": {
              "Ref": "hourlyDataPipelineStateMachine3242CCD2",
            },
            "Id": "Target0",
            "RoleArn": {
              "Fn::GetAtt": [
                "hourlyDataPipelineEventsRuleRole8DA4226F",
                "Arn",
              ],
            },
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
    "hourlyDataPipelineEventsRuleRole8DA4226F": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "events.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "hourlyDataPipelineEventsRuleRoleDefaultPolicy3D321940": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "states:StartExecution",
              "Effect": "Allow",
              "Resource": {
                "Ref": "hourlyDataPipelineStateMachine3242CCD2",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "hourlyDataPipelineEventsRuleRoleDefaultPolicy3D321940",
        "Roles": [
          {
            "Ref": "hourlyDataPipelineEventsRuleRole8DA4226F",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "hourlyDataPipelineExecutionAbortedAlarmCA2416F3": {
      "Properties": {
        "AlarmDescription": "Alarm for the number of executions that aborted exceeded the threshold of 1. ",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "StateMachineArn",
            "Value": {
              "Ref": "hourlyDataPipelineStateMachine3242CCD2",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ExecutionsAborted",
        "Namespace": "AWS/States",
        "Period": 300,
        "Statistic": "Maximum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "hourlyDataPipelineExecutionFailedAlarmC1E4E942": {
      "Properties": {
        "AlarmDescription": "Alarm for the number of executions that failed exceeded the threshold of 1. ",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "StateMachineArn",
            "Value": {
              "Ref": "hourlyDataPipelineStateMachine3242CCD2",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ExecutionsFailed",
        "Namespace": "AWS/States",
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "hourlyDataPipelineExecutionThrottledAlarmA4DB0E83": {
      "Properties": {
        "AlarmDescription": "Alarm for the number of executions that throttled exceeded the threshold of 1. ",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "StateMachineArn",
            "Value": {
              "Ref": "hourlyDataPipelineStateMachine3242CCD2",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "ExecutionThrottled",
        "Namespace": "AWS/States",
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "hourlyDataPipelineStateMachine3242CCD2": {
      "DeletionPolicy": "Delete",
      "DependsOn": [
        "hourlyDataPipelineStateMachineRoleDefaultPolicyF00A5B3B",
        "hourlyDataPipelineStateMachineRole1832C3AF",
      ],
      "Properties": {
        "DefinitionString": {
          "Fn::Join": [
            "",
            [
              "{"StartAt":"LambdaA","States":{"LambdaA":{"Next":"Results Empty?","Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Type":"Task","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "lambdaA49948E53",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"Results Empty?":{"Type":"Choice","InputPath":"$.Payload","Choices":[{"Variable":"$.results","BooleanEquals":false,"Next":"LambdaA"},{"Variable":"$.results","BooleanEquals":true,"Next":"LambdaBMap"}]},"LambdaBMap":{"Type":"Map","End":true,"ItemsPath":"$.orders","ItemProcessor":{"ProcessorConfig":{"Mode":"INLINE"},"StartAt":"LambdaB","States":{"LambdaB":{"End":true,"Retry":[{"ErrorEquals":["Lambda.ClientExecutionTimeoutException","Lambda.ServiceException","Lambda.AWSLambdaException","Lambda.SdkClientException"],"IntervalSeconds":2,"MaxAttempts":6,"BackoffRate":2}],"Catch":[{"ErrorEquals":["States.TaskFailed"],"Next":"Publish Error"}],"Type":"Task","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke","Parameters":{"FunctionName":"",
              {
                "Fn::GetAtt": [
                  "lambdaBE6265022",
                  "Arn",
                ],
              },
              "","Payload.$":"$"}},"Publish Error":{"End":true,"Type":"Task","Resource":"arn:",
              {
                "Ref": "AWS::Partition",
              },
              ":states:::sns:publish","Parameters":{"TopicArn":"",
              {
                "Ref": "ErrorTopicA0904A23",
              },
              "","Message.$":"$"}}}}}}}",
            ],
          ],
        },
        "LoggingConfiguration": {
          "Destinations": [
            {
              "CloudWatchLogsLogGroup": {
                "LogGroupArn": {
                  "Fn::GetAtt": [
                    "hourlyDataPipelineStateMachineLogGroup6C3013B1",
                    "Arn",
                  ],
                },
              },
            },
          ],
          "Level": "ERROR",
        },
        "RoleArn": {
          "Fn::GetAtt": [
            "hourlyDataPipelineStateMachineRole1832C3AF",
            "Arn",
          ],
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
      "UpdateReplacePolicy": "Delete",
    },
    "hourlyDataPipelineStateMachineLogGroup6C3013B1": {
      "DeletionPolicy": "Retain",
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W86",
              "reason": "Retention period for CloudWatchLogs LogGroups are set to 'Never Expire' to preserve customer data indefinitely",
            },
            {
              "id": "W84",
              "reason": "By default CloudWatchLogs LogGroups data is encrypted using the CloudWatch server-side encryption keys (AWS Managed Keys)",
            },
          ],
        },
      },
      "Properties": {
        "LogGroupName": {
          "Fn::Join": [
            "",
            [
              "/aws/vendedlogs/states/constructs/EntrixDataPipelinehourlyDataPipelineStateMachineLog-",
              {
                "Fn::Select": [
                  2,
                  {
                    "Fn::Split": [
                      "/",
                      {
                        "Ref": "AWS::StackId",
                      },
                    ],
                  },
                ],
              },
            ],
          ],
        },
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Retain",
    },
    "hourlyDataPipelineStateMachineRole1832C3AF": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": {
                  "Fn::FindInMap": [
                    "ServiceprincipalMap",
                    {
                      "Ref": "AWS::Region",
                    },
                    "states",
                  ],
                },
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "hourlyDataPipelineStateMachineRoleDefaultPolicyF00A5B3B": {
      "Metadata": {
        "cfn_nag": {
          "rules_to_suppress": [
            {
              "id": "W12",
              "reason": "These are CDK defaults. The 'LogDelivery' actions do not support resource-level authorizations. Any logging is done by State Machine code",
            },
          ],
        },
      },
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "lambdaA49948E53",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "lambdaA49948E53",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": [
                {
                  "Fn::GetAtt": [
                    "lambdaBE6265022",
                    "Arn",
                  ],
                },
                {
                  "Fn::Join": [
                    "",
                    [
                      {
                        "Fn::GetAtt": [
                          "lambdaBE6265022",
                          "Arn",
                        ],
                      },
                      ":*",
                    ],
                  ],
                },
              ],
            },
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Ref": "ErrorTopicA0904A23",
              },
            },
            {
              "Action": [
                "logs:CreateLogDelivery",
                "logs:GetLogDelivery",
                "logs:UpdateLogDelivery",
                "logs:DeleteLogDelivery",
                "logs:ListLogDeliveries",
                "logs:PutResourcePolicy",
                "logs:DescribeResourcePolicies",
                "logs:DescribeLogGroups",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "hourlyDataPipelineStateMachineRoleDefaultPolicyF00A5B3B",
        "Roles": [
          {
            "Ref": "hourlyDataPipelineStateMachineRole1832C3AF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "lambdaA49948E53": {
      "DependsOn": [
        "lambdaAServiceRoleE538E259",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "44622ce00c3c6b4fca5495ae4fb89a5878c0288cadc79326428ffb1f8cd7fd40.zip",
        },
        "Handler": "lambda_a.lambda_handler",
        "Role": {
          "Fn::GetAtt": [
            "lambdaAServiceRoleE538E259",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaAServiceRoleE538E259": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdaBE6265022": {
      "DependsOn": [
        "lambdaBServiceRoleDefaultPolicy3908E73A",
        "lambdaBServiceRole239D9FB4",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "44622ce00c3c6b4fca5495ae4fb89a5878c0288cadc79326428ffb1f8cd7fd40.zip",
        },
        "Environment": {
          "Variables": {
            "LOG_BUCKET": {
              "Ref": "MyLogBucketBD20E7D7",
            },
          },
        },
        "Handler": "lambda_b.lambda_handler",
        "Role": {
          "Fn::GetAtt": [
            "lambdaBServiceRole239D9FB4",
            "Arn",
          ],
        },
        "Runtime": "python3.9",
      },
      "Type": "AWS::Lambda::Function",
    },
    "lambdaBServiceRole239D9FB4": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "lambdaBServiceRoleDefaultPolicy3908E73A": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "s3:PutObject",
                "s3:PutObjectLegalHold",
                "s3:PutObjectRetention",
                "s3:PutObjectTagging",
                "s3:PutObjectVersionTagging",
                "s3:Abort*",
              ],
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    {
                      "Fn::GetAtt": [
                        "MyLogBucketBD20E7D7",
                        "Arn",
                      ],
                    },
                    "/*",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "lambdaBServiceRoleDefaultPolicy3908E73A",
        "Roles": [
          {
            "Ref": "lambdaBServiceRole239D9FB4",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
